src/client/meshgen/extruder.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+#include <array>
+#include <vector>
+#include "irrlichttypes_bloated.h"
+#include <ITexture.h>
+#include <S3DVertex.h>
+
+struct ExtrudedMesh
+{
+	std::vector<video::S3DVertex> vertices;
+	std::vector<u16> indices;
+	std::vector<video::S3DVertex> overlay_vertices;
+	std::vector<u16> overlay_indices;
+
+	explicit operator bool() const noexcept { return !vertices.empty(); }
+};
+
+/*
+ * Handles conversion of textures to extruded meshes.
+ * Each instance may only be used once.
+ */
+class Extruder
+{
+	enum FaceDir
+	{
+		Up,
+		Down,
+		Left,
+		Right,
+	};
+
+	static constexpr int threshold = 128;
+	static constexpr float thickness = 0.1f;
+	video::ITexture *texture;
+	const video::SColor *data;
```

Comment:
Please initialize `data` (and optionally `texture`) to `nullptr` for safety reasons.

Reply:
what safety? `texture` is set in the ctor’s initializers list, it just may not be not set. `data` is set a bit later, but still in the ctor. I updated it to throw instead of returning in the case of error, so an object without `data` set just can’t exist.